src/typescript-worker/types-fetcher.ts

```diff
@@ -9,244 +9,385 @@ import {ModuleResolver} from './module-resolver.js';
-        // anything special with errors (though we could potentially surface
-        // more information).
-        files.set(specifier, fetched.result);
-      }
-    }
-    return files;
+    await this._materializeNodeModulesTree(layout, files, '');
+    // Note in practice we only really need "files", but it's useful to also
+    // return the dependency graph and layout for testing.
+    return {
+      files,
+      layout,
+      dependencyGraph: {
+        root: this._rootDependencies,
+        deps: this._dependencyGraph,
+      },
+    };
   }
 
-  private async _addLibTypings(lib: string): Promise<void> {
-    await this._handleBareSpecifier(
-      `typescript/lib/lib.${lib.toLowerCase()}.js`
+  private async _addTypeScriptStandardLib(lib: string): Promise<void> {
+    return this._handleBareSpecifier(
+      `typescript/lib/lib.${lib.toLowerCase()}.js`,
+      root
     );
   }
 
   private async _handleBareAndRelativeSpecifiers(
     sourceText: string,
-    referrerSpecifier: NodePackage
+    referrer: NpmFileLocation | typeof root
   ): Promise<void> {
     const fileInfo = ts.preProcessFile(sourceText, undefined, true);
     const promises = [];
     for (const {fileName: specifier} of fileInfo.importedFiles) {
       const kind = classifySpecifier(specifier);
       if (kind === 'bare') {
-        promises.push(this._handleBareSpecifier(specifier));
-      } else if (kind === 'relative') {
-        promises.push(
-          this._handleRelativeSpecifier(specifier, referrerSpecifier)
-        );
+        promises.push(this._handleBareSpecifier(specifier, referrer));
+      } else if (kind === 'relative' && referrer !== root) {
+        // Note we never need to follow relative imports from project root files
+        // because those can only be other project files, which are already
+        // being processed, since we pass all project files to this class.
+        promises.push(this._handleRelativeSpecifier(specifier, referrer));
       }
     }
     for (const {fileName: lib} of fileInfo.libReferenceDirectives) {
-      promises.push(this.addLibTypings(lib));
+      promises.push(this._addTypeScriptStandardLib(lib));
     }
-    await Promise.all(promises);
+    await Promise.allSettled(promises);
   }
 
-  private async _handleBareSpecifier(bare: string): Promise<void> {
-    if (this._handledSpecifiers.has(bare)) {
+  private async _handleBareSpecifier(
+    bare: string,
+    referrer: NpmFileLocation | typeof root
+  ): Promise<void> {
+    let location = parseNpmStyleSpecifier(bare);
+    if (location === undefined) {
       return;
     }
-    this._handledSpecifiers.add(bare);
-    const npm = parseNpmStyleSpecifier(bare);
-    if (npm === undefined) {
-      return;
+    // Versions don't make much sense with an import map, since you just have
+    // bare specifier -> URL. We can leave whatever version we already have; it
+    // will always be ignored.
```

Comment:
If there are two different imports in the import map that reference the same url will the work be duplicated and aliased under the two different importMap keys?

I.e.:
```
"imports": {
    "a": "<URL_TO_A>",
    "b": "<URL_TO_A>",
}
```

Reply:
This works -- the typings will appear both under `node_modules/a` and `node_modules/b`. I updated a test case to confirm.

As written, there is actually a duplicate fetch, because our fetch cache is currently keyed by specifier, not URL. However, I think I'm pretty ok with ignoring that because 1. Mapping two packages to the same URL seems really ununsual, and 2. The browser HTTP cache should still get a hit, just a little deeper in the stack.